tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AWS/SQS/AwsSqsHandlerCommon.cs

```diff
@@ -0,0 +1,170 @@
+// <copyright file="AwsSqsHandlerCommon.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using Datadog.Trace.ClrProfiler.CallTarget;
+using Datadog.Trace.DataStreamsMonitoring;
+using Datadog.Trace.DuckTyping;
+using Datadog.Trace.Vendors.Newtonsoft.Json.Utilities;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.AWS.SQS;
+
+/// <summary>
+/// Contains the code that is shared between the integrations of sync/async and batch/single send and receive.
+/// </summary>
+internal static class AwsSqsHandlerCommon
+{
+    internal static CallTargetState BeforeSend<TSendMessageRequest>(TSendMessageRequest request, SendType sendType)
+    {
+        if (request is null)
+        {
+            return CallTargetState.GetDefault();
+        }
+
+        // we can't use generic constraints for this duck typing, because we need the original type
+        // for the InjectHeadersIntoMessage<TSendMessageRequest> call below
+        var requestProxy = request.DuckCast<IAmazonSQSRequestWithQueueUrl>();
+
+        var scope = AwsSqsCommon.CreateScope(Tracer.Instance, sendType.OperationName, out var tags, spanKind: SpanKinds.Producer);
+
+        var queueName = AwsSqsCommon.GetQueueName(requestProxy.QueueUrl);
+        if (tags is not null && requestProxy.QueueUrl is not null)
+        {
+            tags.QueueUrl = requestProxy.QueueUrl;
+            tags.QueueName = queueName;
+        }
+
+        if (scope?.Span.Context != null && !string.IsNullOrEmpty(queueName))
+        {
+            var dataStreamsManager = Tracer.Instance.TracerManager.DataStreamsManager;
```

Comment:
I _think_ this _can_ be `null`, based on your other checks elsewhere? If so, we need to check for that in `InjectForSingleMessage` and `InjectForBatch`